lua/nvim-surround/init.lua

```diff
@@ -195,6 +193,34 @@ M.change_surround = function(args)
     local selections = utils.get_nearest_selections(args.del_char, "change")
     local delimiters = args.add_delimiters()
     if selections and delimiters then
+        -- Avoid adding any, and remove any existing whitespace after the
+        -- opening delimiter if only whitespace exists between it and the end
+        -- of the line. Avoid adding or removing leading whitespace before the
+        -- closing delimiter if only whitespace exists between it and the
+        -- beginning of the line.
+
+        -- Column number of end of opening selection
+        local xpos = selections.left.last_pos[2]
+        local space_begin, space_end = buffer.get_line(selections.left.last_pos[1]):find("%s*$")
+
+        if space_begin - 1 <= xpos then -- Whitespace is adjacent to opening delimiter
```

Comment:
Would a better heuristic here be to check if the line is entirely whitespace, i.e.
```lua
if space_begin == 1 then
    -- Continue logic...
```

Reply:
Unless we're matching a delimiter that is itself nothing but whitespace, `space_begin` should never be `1` in this case. Likewise when matching leading whitespace on the closing delimiter's line, `space_end` should always be an index less than the last column.

I didn't exactly write the most intuitive code here, but the logic is to test for contiguous whitespace between the opening delimiter and the end of its line, and between the closing delimiter and the beginning of its line. If the delimiters contain whitespace (which is the only case we actually need to do anything) then `find()` will match that whitespace too, hence the `<=` and `>=` rather than `==`.

There's no logic relating to moving the position of either delimiter, only adjusting whether the replacement delimiters include leading/trailing whitespace when they are inserted, and adjusting the regions that get removed before those replacements are inserted.

The replacement region for the opening delimiter gets expanded to include all following whitespace until the end of the line, if there is nothing else between it and the end of the line, so any existing trailing whitespace gets removed. The region for the closing delimiter gets shrunk to exclude all whitespace between it and the beginning of the line, if there is nothing else between it and the beginning of the line, so any existing leading whitespace is left unchanged so as to preserve whatever indent the line has.